util/udev/qmk_id.c

```diff
@@ -0,0 +1,160 @@
+/*
```

Comment:
Could this be done in Python, in the CLI, instead?

Reply:
This will be run by udev every time any hidraw device is plugged in.
It should really be as fast and dependency-free as possible.

Reply:
Also it will be run as the root user. It would probably be good to also drop all privileges to protect against issues in the hid parser.

Reply:
I would strongly prefer our python code not be run as root. :)

Reply:
Sounds reasonable :-)

I'll add some additional hardening.

Reply:
Having some random C code that needs to be compiled and run separately in order to get this working just seems to me a little messy, is all.

Reply:
Yeah, this is fairly peculiar.
The alternatives would be:
* Open access to all hidraw devices
   * (This is the current behavior)
   * Also grants access to devices that have nothing to do with QMK
* Provide a static list of all devices that are using QMK
   * This will be a long list
   * Everything not part of QMK mainline will need to add their own rules

These random C programs are actually how udev is supposed to be extended for more dynamic logic.
On my machine I have also `ata_id`  `cdrom_id`  `dmi_memory_id`  `fido_id`  `kpartx_id` `scsi_id` and `v4l_id`.
From a Linux distribution point I think it would be better keep the installed rules as specific as possible and they should be able to nicely install this as part of their packages.

Reply:
Now the sandbox should be fairly tight.

Reply:
> Having some random C code that needs to be compiled and run separately in order to get this working just seems to me a little messy, is all.

Since this is a bit beyond me....

Is the C code compiled each time it's ran?  Or do you need to run it once and that's mostly it? 
If it's just the once, would it make sense to add it to the setup routine for linux, so it's "ready to go"?

Reply:
No, it is only compiled once and then installed together with the udev rules in a system-wide directory.

> would it make sense to add it to the setup routine for linux

Yes, good point. I'll add it there.

Reply:
@drashna Currently the udev rules file is not installed through the setup routine but setup instructions are shown to the user.
The rules file and `qmk_id` only work together, so they should be handled the same.
Should both be installed automatically or they both be mentioned in the message?

Reply:
Also currently the installer looks at the actually installed `/etc/udev/rules.d/50-qmk.rules` and validates it, having a complete duplication of the rules files in python again.
Then it prompts the user to install the rules to `/etc/udev/rules.d` manually.

Instead the installer could just drop the rules file itself into `/usr/{local,}/udev/rules,d/50-qmk.rules` and then ask the user to delete their `/etc/udev/rules.d/50-qmk.rules` because that name is now managed by the installer exclusively.
Instead they can put their customization into `49-qmk-custom.rules` for example.
(Or the installer uses a different name that does not conflict with the existing name)

This is mostly a policy decision on how the installer is supposed to behave. To be honest I have no experience with the installer as I'm using the ArchLinux package.

FYI these are the semantics of the different locations (from `man 7 udev`):

The udev rules are read from the files located in the system rules directories /usr/lib/udev/rules.d and /usr/local/lib/udev/rules.d, the volatile runtime directory /run/udev/rules.d and the local administration directory /etc/udev/rules.d. All rules files are collectively sorted and processed in lexical order, regardless of the directories in which they live. However, files with identical filenames replace each other. Files in /etc/ have the highest priority, files in /run/ take precedence over files with the same name under /usr/. This can be used to override a system-supplied rules file with a local file if needed; a symlink in /etc/ with the same name as a rules file in /usr/lib/, pointing to /dev/null, disables the rules file entirely. Rule files must have the extension .rules; other extensions are ignored.

Reply:
@drashna I think we still need to decide how the installation should happen.
(And then implement that)

Reply:
> having a complete duplication of the rules files in python again

This is a necessary evil. The dedicated file was requested by distro maintainers (I think Arch) for easy packaging.

> Then it prompts the user to install the rules to /etc/udev/rules.d manually.

This is something that we've decided to leave up to the user. There are users who do not like udev for a reason or another. There are also multiple udev implementations with varying degree of compatibility.

Reply:
@Erovia  So should the user also be prompted to install the `qmk_id` tool on their own?
As it stands the udev rules in this PR do need the tool to properly handle `qmk console` (without opening all devices to everyone).